File: days/day_08/solutions/solution.ts
import * as fs from 'fs';
import * as path from 'path';

type Coord = [number, number, number];
type Pair = [number, number, number]; // [i, j, dist_sq]

class UnionFind {
    private parent: number[];
    private size: number[];
    public componentCount: number;

    constructor(n: number) {
        this.parent = Array.from({ length: n }, (_, i) => i);
        this.size = Array(n).fill(1);
        this.componentCount = n;
    }

    find(x: number): number {
        if (this.parent[x] !== x) {
            this.parent[x] = this.find(this.parent[x]);
        }
        return this.parent[x];
    }

    union(x: number, y: number): boolean {
        let rootX = this.find(x);
        let rootY = this.find(y);
        if (rootX === rootY) {
            return false;
        }
        if (this.size[rootX] < this.size[rootY]) {
            [rootX, rootY] = [rootY, rootX];
        }
        this.parent[rootY] = rootX;
        this.size[rootX] += this.size[rootY];
        this.componentCount--;
        return true;
    }
}

function parseCoordinates(lines: string[]): Coord[] {
    const coords: Coord[] = [];
    for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed) continue;
        const parts = trimmed.split(',');
        const x = parseInt(parts[0], 10);
        const y = parseInt(parts[1], 10);
        const z = parseInt(parts[2], 10);
        coords.push([x, y, z]);
    }
    return coords;
}

function squaredDistance(p1: Coord, p2: Coord): number {
    const dx = p2[0] - p1[0];
    const dy = p2[1] - p1[1];
    const dz = p2[2] - p1[2];
    return dx * dx + dy * dy + dz * dz;
}

function solve(inputData: string): [string, string] {
    const lines = inputData.trim().split('\n');
    const coords = parseCoordinates(lines);
    
    const n = coords.length;
    if (n === 0) {
        return ["0", "0"];
    }
    
    // Generate all pairs with squared distances
    const pairs: Pair[] = [];
    for (let i = 0; i < n; i++) {
        for (let j = i + 1; j < n; j++) {
            const distSq = squaredDistance(coords[i], coords[j]);
            pairs.push([i, j, distSq]);
        }
    }
    
    // Sort by distance
    pairs.sort((a, b) => a[2] - b[2]);
    
    // Part 1: Connect first 1000 pairs
    const uf1 = new UnionFind(n);
    let connectionsMade = 0;
    for (const [i, j] of pairs) {
        if (connectionsMade >= 1000) break;
        uf1.union(i, j);
        connectionsMade++;
    }
    
    // Get component sizes
    const componentSizes: Map<number, number> = new Map();
    for (let i = 0; i < n; i++) {
        const root = uf1.find(i);
        componentSizes.set(root, uf1['size'][root]);
    }
    
    const sizes = Array.from(componentSizes.values()).sort((a, b) => b - a);
    const part1 = sizes.length >= 3 ? sizes[0] * sizes[1] * sizes[2] : 0;
    
    // Part 2: Connect until all in one circuit
    const uf2 = new UnionFind(n);
    let finalPair: [number, number] | null = null;
    for (const [i, j] of pairs) {
        if (uf2.componentCount === 1) break;
        if (uf2.union(i, j)) {
            if (uf2.componentCount === 1) {
                finalPair = [i, j];
                break;
            }
        }
    }
    
    let part2 = 0;
    if (finalPair) {
        const [i, j] = finalPair;
        part2 = coords[i][0] * coords[j][0];
    }
    
    return [String(part1), String(part2)];
}

function main(): void {
    const inputPath = path.join(__dirname, '../data/input.txt');
    const data = fs.readFileSync(inputPath, 'utf-8');
    const [part1, part2] = solve(data);
    console.log(`Part 1: ${part1}`);
    console.log(`Part 2: ${part2}`);
}

main();

File: days/day_08/solutions/solution.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

class Solution {
    static class Coord {
        long x, y, z;
        Coord(long x, long y, long z) {
            this.x = x; this.y = y; this.z = z;
        }
    }
    
    static class Pair {
        int i, j;
        long distSq;
        Pair(int i, int j, long distSq) {
            this.i = i; this.j = j; this.distSq = distSq;
        }
    }
    
    static class UnionFind {
        int[] parent;
        long[] size;
        int componentCount;
        
        UnionFind(int n) {
            parent = new int[n];
            size = new long[n];
            for (int i = 0; i < n; i++) {
                parent[i] = i;
                size[i] = 1;
            }
            componentCount = n;
        }
        
        int find(int x) {
            if (parent[x] != x) {
                parent[x] = find(parent[x]);
            }
            return parent[x];
        }
        
        boolean union(int x, int y) {
            int rootX = find(x);
            int rootY = find(y);
            if (rootX == rootY) {
                return false;
            }
            if (size[rootX] < size[rootY]) {
                int temp = rootX;
                rootX = rootY;
                rootY = temp;
            }
            parent[rootY] = rootX;
            size[rootX] += size[rootY];
            componentCount--;
            return true;
        }
    }
    
    static List<Coord> parseCoordinates(String[] lines) {
        List<Coord> coords = new ArrayList<>();
        for (String line : lines) {
            line = line.trim();
            if (line.isEmpty()) continue;
            String[] parts = line.split(",");
            long x = Long.parseLong(parts[0]);
            long y = Long.parseLong(parts[1]);
            long z = Long.parseLong(parts[2]);
            coords.add(new Coord(x, y, z));
        }
        return coords;
    }
    
    static long squaredDistance(Coord p1, Coord p2) {
        long dx = p2.x - p1.x;
        long dy = p2.y - p1.y;
        long dz = p2.z - p1.z;
        return dx * dx + dy * dy + dz * dz;
    }
    
    public static String[] solve(String inputData) {
        String[] lines = inputData.trim().split("\n");
        List<Coord> coords = parseCoordinates(lines);
        
        int n = coords.size();
        if (n == 0) {
            return new String[]{"0", "0"};
        }
        
        // Generate all pairs with squared distances
        List<Pair> pairs = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                long distSq = squaredDistance(coords.get(i), coords.get(j));
                pairs.add(new Pair(i, j, distSq));
            }
        }
        
        // Sort by distance
        pairs.sort((a, b) -> Long.compare(a.distSq, b.distSq));
        
        // Part 1: Connect first 1000 pairs
        UnionFind uf1 = new UnionFind(n);
        int connectionsMade = 0;
        for (Pair pair : pairs) {
            if (connectionsMade >= 1000) break;
            uf1.union(pair.i, pair.j);
            connectionsMade++;
        }
        
        // Get component sizes
        Map<Integer, Long> componentSizes = new HashMap<>();
        for (int i = 0; i < n; i++) {
            int root = uf1.find(i);
            componentSizes.put(root, uf1.size[root]);
        }
        
        List<Long> sizes = new ArrayList<>(componentSizes.values());
        sizes.sort((a, b) -> Long.compare(b, a));
        long part1 = sizes.size() >= 3 ? sizes.get(0) * sizes.get(1) * sizes.get(2) : 0;
        
        // Part 2: Connect until all in one circuit
        UnionFind uf2 = new UnionFind(n);
        Pair finalPair = null;
        for (Pair pair : pairs) {
            if (uf2.componentCount == 1) break;
            if (uf2.union(pair.i, pair.j)) {
                if (uf2.componentCount == 1) {
                    finalPair = pair;
                    break;
                }
            }
        }
        
        long part2 = 0;
        if (finalPair != null) {
            part2 = coords.get(finalPair.i).x * coords.get(finalPair.j).x;
        }
        
        return new String[]{String.valueOf(part1), String.valueOf(part2)};
    }
    
    public static void main(String[] args) throws IOException {
        Path inputPath = Paths.get("../data/input.txt");
        String data = Files.readString(inputPath);
        String[] results = solve(data);
        System.out.println("Part 1: " + results[0]);
        System.out.println("Part 2: " + results[1]);
    }
}
